sklearn/utils/_param_validation.py

```diff
@@ -249,6 +251,31 @@ def __str__(self):
         return "None"
 
 
+class _NanConstraint(_Constraint):
+    """Constraint representing the indicator `np.nan`."""
+
+    def is_satisfied_by(self, val):
+        return isinstance(val, Real) and np.isnan(val)
```

Comment:
Nit: Given this only works on scalars:

```suggestion
        return isinstance(val, Real) and math.isnan(val)
```

Also do not need to go through all the NumPy machinery:

```python
%%timeit
math.isnan(np.nan)
# 60.9 ns ± 0.0996 ns per loop

%%timeit
np.isnan(np.nan)
$ 531 ns ± 2.25 ns per loop
```